BUG: special tests failing on main when CuPy is installed with "wrong" NumPy

On latest `main` on x86_64 Linux on a machine with `cupy` installed `python dev.py test -j 32` produces 32 test errors from `scipy/special/tests/test_xsf_cuda.py`. Since there's no opt-in to array API or anything like that, this can't possibly be the desired behavior. CuPy is `13.1.0`, NumPy is `2.0.0`.

I realize there's some CuPy-specific stuff happening here, but you can't just assume that the version of CuPy someone has installed is compatible with the version of NumPy they have installed. I often swap between `2.x` and `1.x` NumPy versions during development/array API work for example, and needing to swap CuPy versions as well would be a bit painful.

A sample traceback is below the fold:

<details>

```________________________________________________ ERROR collecting build-install/lib/python3.11/site-packages/scipy/special/tests/test_xsf_cuda.py _________________________________________________
scipy/special/tests/test_xsf_cuda.py:13: in <module>
    import cupy  # type: ignore
        MissingModule = <class 'scipy.special._testutils.MissingModule'>
        __builtins__ = <builtins>
        __cached__ = '/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/special/tests/__pycache__/test_xsf_cuda.cpython-311.pyc'
        __doc__    = None
        __file__   = '/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/special/tests/test_xsf_cuda.py'
        __loader__ = <_pytest.assertion.rewrite.AssertionRewritingHook object at 0x7f050a9a0510>
        __name__   = 'scipy.special.tests.test_xsf_cuda'
        __package__ = 'scipy.special.tests'
        __spec__   = ModuleSpec(name='scipy.special.tests.test_xsf_cuda', loader=<_pytest.assertion.rewrite.AssertionRewritingHook object a...n='/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/special/tests/test_xsf_cuda.py')
        check_version = <function check_version at 0x7f04a274b1a0>
        os         = <module 'os' (frozen)>
        pytest     = <module 'pytest' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/pytest/__init__.py'>
        sc         = <module 'scipy.special' from '/home/treddy/github_projects/scipy/build-install/lib/python3.11/site-packages/scipy/special/__init__.py'>
        shutil     = <module 'shutil' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/shutil.py'>
        tempfile   = <module 'tempfile' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/tempfile.py'>
        uuid4      = <function uuid4 at 0x7f050abf0d60>
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py:47: in <module>
    from cupy import testing  # NOQA  # NOQA
        __builtins__ = <builtins>
        __cached__ = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__pycache__/__init__.cpython-311.pyc'
        __doc__    = None
        __file__   = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f04a02ab4d0>
        __name__   = 'cupy'
        __package__ = 'cupy'
        __path__   = ['/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy']
        __spec__   = ModuleSpec(name='cupy', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f04a02ab4d0>, origin='/home/t...nit__.py', submodule_search_locations=['/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy'])
        __version__ = '13.1.0'
        _core      = <module 'cupy._core' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_core/__init__.py'>
        _creation  = <module 'cupy._creation' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_creation/__init__.py'>
        _cupyx     = <module 'cupyx' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupyx/__init__.py'>
        _environment = <module 'cupy._environment' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_environment.py'>
        _functools = <module 'functools' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/functools.py'>
        _numpy     = <module 'numpy' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/numpy/__init__.py'>
        _sys       = <module 'sys' (built-in)>
        _util      = <module 'cupy._util' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_util.cpython-311-x86_64-linux-gnu.so'>
        _version   = <module 'cupy._version' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/_version.py'>
        cuda       = <module 'cupy.cuda' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/cuda/__init__.py'>
        fft        = <module 'cupy.fft' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/fft/__init__.py'>
        is_available = <function is_available at 0x7f04a0337380>
        linalg     = <module 'cupy.linalg' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/linalg/__init__.py'>
        polynomial = <module 'cupy.polynomial' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/polynomial/__init__.py'>
        random     = <module 'cupy.random' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/random/__init__.py'>
        sparse     = <module 'cupy.sparse' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/sparse/__init__.py'>
        typing     = <module 'cupy.typing' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/typing/__init__.py'>
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/__init__.py:22: in <module>
    from cupy.testing._loops import for_all_dtypes  # NOQA
        AssertFunctionIsCalled = <function assert_function_is_called at 0x7f049fdb28e0>
        NumpyAliasBasicTestBase = <class 'cupy.testing._helper.NumpyAliasBasicTestBase'>
        NumpyAliasValuesTestBase = <class 'cupy.testing._helper.NumpyAliasValuesTestBase'>
        __builtins__ = <builtins>
        __cached__ = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/__pycache__/__init__.cpython-311.pyc'
        __doc__    = None
        __file__   = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/__init__.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f049fdb8c10>
        __name__   = 'cupy.testing'
        __package__ = 'cupy.testing'
        __path__   = ['/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing']
        __spec__   = ModuleSpec(name='cupy.testing', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f049fdb8c10>, origin=...', submodule_search_locations=['/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing'])
        _array     = <module 'cupy.testing._array' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_array.py'>
        _attr      = <module 'cupy.testing._attr' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_attr.py'>
        _bundle    = <module 'cupy.testing._bundle' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_bundle.py'>
        _helper    = <module 'cupy.testing._helper' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_helper.py'>
        _parameterized = <module 'cupy.testing._parameterized' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_parameterized.py'>
        _pytest_impl = <module 'cupy.testing._pytest_impl' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_pytest_impl.py'>
        assert_allclose = <function assert_allclose at 0x7f049fdb0900>
        assert_array_almost_equal = <function assert_array_almost_equal at 0x7f049fdb09a0>
        assert_array_almost_equal_nulp = <function assert_array_almost_equal_nulp at 0x7f049fdb0a40>
        assert_array_equal = <function assert_array_equal at 0x7f049fdb0b80>
        assert_array_less = <function assert_array_less at 0x7f049fdb0cc0>
        assert_array_list_equal = <function assert_array_list_equal at 0x7f049fdb0c20>
        assert_array_max_ulp = <function assert_array_max_ulp at 0x7f049fdb0ae0>
        assert_warns = <function assert_warns at 0x7f049fdb2340>
        generate_matrix = <function generate_matrix at 0x7f049fdb2200>
        installed  = <function installed at 0x7f049fdb1e40>
        multi_gpu  = <function multi_gpu at 0x7f049fdb0fe0>
        numpy_satisfies = <function numpy_satisfies at 0x7f049fdb1ee0>
        shaped_arange = <function shaped_arange at 0x7f049fdb1f80>
        shaped_random = <function shaped_random at 0x7f049fdb20c0>
        shaped_reverse_arange = <function shaped_reverse_arange at 0x7f049fdb2020>
        shaped_sparse_random = <function shaped_sparse_random at 0x7f049fdb2160>
        slow       = <function slow at 0x7f049fdb0e00>
        with_requires = <function with_requires at 0x7f049fdb1da0>
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_loops.py:109: in <module>
    numpy.AxisError, numpy.linalg.LinAlgError,
        Tuple      = typing.Tuple
        Type       = typing.Type
        __annotations__ = {'_skip_classes': typing.Tuple[typing.Type, ...]}
        __builtins__ = <builtins>
        __cached__ = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/__pycache__/_loops.cpython-311.pyc'
        __doc__    = None
        __file__   = '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_loops.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f049fdc4490>
        __name__   = 'cupy.testing._loops'
        __package__ = 'cupy.testing'
        __spec__   = ModuleSpec(name='cupy.testing._loops', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f049fdc4490>, origin='/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_loops.py')
        _array     = <module 'cupy.testing._array' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_array.py'>
        _call_func = <function _call_func at 0x7f049fdb2ca0>
        _call_func_cupy = <function _call_func_cupy at 0x7f049fdb2d40>
        _call_func_numpy = <function _call_func_numpy at 0x7f049fdb2de0>
        _call_func_numpy_cupy = <function _call_func_numpy_cupy at 0x7f049fdb2e80>
        _format_exception = <function _format_exception at 0x7f049fdb2c00>
        _is_pytest_available = True
        _parameterized = <module 'cupy.testing._parameterized' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_parameterized.py'>
        _pytest    = <module '_pytest' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/_pytest/__init__.py'>
        _skip_classes = (<class 'unittest.case.SkipTest'>, <class 'Skipped'>)
        cupy       = <module 'cupy' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py'>
        cupyx      = <module 'cupyx' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupyx/__init__.py'>
        functools  = <module 'functools' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/functools.py'>
        inspect    = <module 'inspect' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/inspect.py'>
        is_available = <function is_available at 0x7f049fdb1080>
        numpy      = <module 'numpy' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/numpy/__init__.py'>
        os         = <module 'os' (frozen)>
        random     = <module 'random' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/random.py'>
        traceback  = <module 'traceback' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/traceback.py'>
        unittest   = <module 'unittest' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/unittest/__init__.py'>
        warnings   = <module 'warnings' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/warnings.py'>
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/numpy/__init__.py:410: in __getattr__
    raise AttributeError("module {!r} has no attribute "
E   AttributeError: module 'numpy' has no attribute 'AxisError'
        attr       = 'AxisError'
        warnings   = <module 'warnings' from '/home/treddy/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/warnings.py'>

```

</details>

This is one example of a possible fix that works locally:

```diff
--- a/scipy/special/tests/test_xsf_cuda.py
+++ b/scipy/special/tests/test_xsf_cuda.py
@@ -11,7 +11,7 @@ from scipy.special._testutils import MissingModule
 
 try:
     import cupy  # type: ignore
-except ImportError:
+except (ImportError, AttributeError):
     cupy = MissingModule('cupy')
```

I think my point is that even if we have nice new CuPy-based testing separate from the array API, we should not error those tests out if we can't safely import CuPy--just skip them in that case I think. If someone opts in to i.e., array API I think that's fine for being stricter on failing with version mismatches on optional libs.

--- a/scipy/special/tests/test_xsf_cuda.py
+++ b/scipy/special/tests/test_xsf_cuda.py
@@ -11,7 +11,7 @@ from scipy.special._testutils import MissingModule
 
 try:
     import cupy  # type: ignore
-except ImportError:
+except (ImportError, AttributeError):
     cupy = MissingModule('cupy')
 
 

